### PR TITLE
Fix: More URL failures

### PR DESCRIPTION
**Pull request by patrickhulce**
More consistently handles host matching of URLs coming from userland.

Addresses #1331

### CODE REVIEW

File: lighthouse-core/lib/url-shim.js [hunk 1]
```diff
@@ -25,4 +25,26 @@
...
+/**
+ * Safely checks if the host of a URL matches a known host with a specified
+ * fallback in case of error.
+ *
+ * @param {string} url
+ * @param {string} host
+ * @param {boolean|function=} fallback
+ * @return {boolean}
+ */
+module.exports.hostMatches = function(url, host, fallback) {
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: @ebidel went ahead and added a helper since 2 became 4

File: lighthouse-core/lib/url-shim.js [hunk 2]
```diff
@@ -25,4 +25,26 @@
...
+  try {
+    return new URL(url).host === host;
+  } catch (err) {
+    let result = fallback;
+    if (typeof fallback === 'function') {
```
Comment (brendankenny) [inline lighthouse-core/lib/url-shim.js]: this boolean/function business is definitely a bit wonky, though it doesn't seem like there's a terribly elegant way of handling the three states without a conditional in each place calling this function. It does feel like this function is trying to be a bit too clever, though, when really all we want is a way to catch bad URLs coming through the js call sites function

**Comment by brendankenny** [comment]
While it's good that #1288 and this PR will prevent future issue storms over malformed URLs, it does mean that we'll never even see future errors if there's yet another way that bad URLs creep in as they'll all be absorbed. In this case it's only for invalid URLs, but I'm somewhat torn over that tradeoff, since we'll be dependent on developers noticing and debugging to see what URL wasn't working

File: lighthouse-core/lib/url-shim.js (hunk 2 above)
Comment (paulirish) [inline lighthouse-core/lib/url-shim.js]: I feel similarly odd about this method signature. 

Here's one proposal: Keep the boolean arg, but enforce it. Then, split the callback into its own optional param.

new method signature:

```ts
  const hostMatches = function (url1: string, url2: string, recoverOnException: boolean, callback?: Function) {
    // enforce the recoverOnException boolean arg
    if (typeof recoverOnException === 'undefined')
      throw new Error('recoverOnException must be provided to hostMatches()');
    
    // ...
  }
```

with usage:
```js
    return URL.hostMatches(err.url, pageHost, true, _ => {
        debugString = 'Lighthouse was unable to determine if some API uses ' +
          `were made by this page. It's possible a Chrome extension` +
          `content script or other eval'd code is calling this API.`;
      });
```

**Comment by brendankenny** [comment]
another option would be to fix this at the source by having the `url` property on these call site objects *only* have valid URLs in them, with a warning emitted (or even an error thrown :) from within `driver.captureFunctionCallSites` if one of the returned call sites doesn't have a valid URL. Important thing would be that a caller could always assume that the returned URLs are URLs, because `captureFunctionCallSites` has already checked, and we wouldn't have to continue to be ultra careful with using it in the future.

(if we still wanted to overload the `url` property to also allow `eval(whatever)` and `<anonymous>` as values, the caller would have to also check `isEval` and `isExtension`, but each one wouldn't have to validate the output of `driver.captureFunctionCallSites`)

**Comment by ebidel** [comment]
Part of the problem is that we're not sure when .url can be invalid. We're catching them case by case right now as bug reports come in. First it was the eval case, then extension content scripts, but there could be other reasons in the future (e.g. `callsite.isSuperLameThingThatBlowsUp`)

**Comment by brendankenny** [comment]
Right, the suggestion was just to centralize the catching of these cases so that anything coming out of `captureFunctionCallSites` is a URL (or from a known set of strings like `<anonymous>`, etc) and have that method handle warning or throwing errors for future cases to turn into URLs (or add to the list of known strings)

**Comment by wardpeet** [comment]
I would suggest using a try catch and the new URL inside the `captureFunctionCallSites`. You can keep eval and all other can be mapped into isValid and filter all valid ones inside the gatherer or inside each audit and save the non valid ones inside extendedinfo an show them somewhere on the report..

**Comment by ebidel** [comment]
Note: `extendedInfo` is the place we show the list of URLs. We currently show valid URLs there as well as  "invalid" ones (`eval(whatever)` and `<anonymous>..` as values) to give code context. We don't want to filter out the later b/c they're still valid user callsites.

> another option would be to fix this at the source by having the url property on these call site objects only have valid URLs in them with a warning emitted (or even an error thrown :) in `driver.captureFunctionCallSite`

We don't want to throw b/c non-valid urls don't necessarily mean errors :)

The right cocktail may be a combo of @brendankenny and @wardpeet's suggestions, where we ensure `.url` is always a valid URL from `driver.captureFunctionCallSite` but set an `.isValid = false` flag if it's not. That would allow downstream audits to check for it, set their own `debugString`, but not have to `new URL()` themselves.

File: lighthouse-core/lib/url-shim.js (hunk 1 above)
Comment (ebidel) [inline lighthouse-core/lib/url-shim.js]: Da powerz of ES6.

**Comment by patrickhulce** [comment]
> The right cocktail may be a combo of @brendankenny and @wardpeet's suggestions, where we ensure .url is always a valid URL from driver.captureFunctionCallSite but set an .isValid = false flag if it's not. That would allow downstream audits to check for it, set their own debugString, but not have to new URL() themselves.

I agree on not throwing for all of these and like the idea of `.isValid` flag. The errors this PR is addressing don't come from the `captureFunctionCallSite` results though, so I still think the subclassed `URL` has some value.


> it does mean that we'll never even see future errors if there's yet another way that bad URLs creep in as they'll all be absorbed. In this case it's only for invalid URLs, but I'm somewhat torn over that tradeoff, since we'll be dependent on developers noticing and debugging to see what URL wasn't working

Do we have any sort of strategy for error-tracking these sorts of things instead (using Rollbar or some similar open source tool)? It would be helpful to not have to throw fatal errors just to find out what our corner cases are across the board.

**Comment by brendankenny** [comment]
> The errors this PR is addressing don't come from the captureFunctionCallSite results though

Right, this would move the errors into there, making sure any URLs it says are valid are parseable and valid so each caller of `driver.captureFunctionCallSite` doesn't have to try/catch a parse of the URL

**Comment by patrickhulce** [comment]
No @brendankenny `no-mutation-events` and `uses-passive-event-listeners` don't use any result of `captureFunctionCallSite` at all. They're operating on the dump of event listeners which have their own URLs to fix, which is why I still find value in sharing that functionality.

**Comment by brendankenny** [comment]
oh, whoops, didn't even notice that, sorry :)

I still think the check of if try/catch works on `new URL(url)` should be in `captureFunctionCallSite` and in the `EventListeners` gatherers, I guess. I'm just not wild about these URL foot guns hanging around, but we can defer if you want.

**Comment by brendankenny** [comment]
> Do we have any sort of strategy for error-tracking these sorts of things instead (using Rollbar or some similar open source tool)? It would be helpful to not have to throw fatal errors just to find out what our corner cases are across the board.

Short of app analytics (which I'm generally not a fan of) I think we talked about adding something similar to the "Report Issue" link in the extension but to the top of the report itself, saying something like "here's your results, but there was an error encountered while running, click this link to report the issue". This would work for the CLI as well, but it would require looking at the report and clicking the link, so it wouldn't really work for anyone using LH in CI or whatever. Not sure if there were other ideas.

**Comment by patrickhulce** [comment]
OK I took a step back from fanciness of a do-it-all function and just cleaned up the try/catch. Even if there's some duplication IMO it's a lot easier to understand than a magic function, but open to other approaches if it isn't that confusing afterall.

File: lighthouse-core/lib/url-shim.js (hunk 2 above)
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: Yeah I agree it feels a little forced. Switched to simpler separation entirely, lmk if you still prefer this approach though

File: lighthouse-core/lib/url-shim.js [hunk 3]
```diff
@@ -25,4 +25,26 @@
...
 // https://github.com/GoogleChrome/lighthouse/issues/1186
-module.exports = (typeof self !== 'undefined' && self.URL) || require('whatwg-url').URL;
+const URL = module.exports = (typeof self !== 'undefined' && self.URL) || require('whatwg-url').URL;
+
+/**
+ * Safely checks if the host of a URL matches a known host with a specified
+ * fallback in case of error.
+ *
+ * @param {string} url
+ * @param {string} host
```
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: done

File: lighthouse-core/lib/url-shim.js (hunk 1 above)
Comment (patrickhulce) [inline lighthouse-core/lib/url-shim.js]: done

**Pull request merged**